File: intellij-plugin/src/main/kotlin/dev/skillbill/intellij/domain/LastKnownDisplayCache.kt
package dev.skillbill.intellij.domain

import java.time.Instant

/**
 * Last-known display snapshot suitable for preference persistence.
 * Observation time is required whenever a cached snapshot exists.
 * Consumers may surface this only as [SkillBillStatusOutcome.Stale], never as
 * an authoritative active state.
 */
data class LastKnownDisplayCache(
    val display: CachedDisplaySnapshot,
    val observedAt: Instant,
)

/**
 * Safe, bounded fields for stale UI fallback. No tokens, prompts, stderr,
 * absolute paths, or unbounded blobs.
 */
data class CachedDisplaySnapshot(
    val summary: String,
    val repositoryIdentity: String? = null,
    val issueKey: String? = null,
    val currentStepId: String? = null,
    val currentStepLabel: String? = null,
    val progressCompleted: Int? = null,
    val progressTotal: Int? = null,
    val startedAt: Instant? = null,
    val currentSubtaskId: String? = null,
    val subtaskStartedAt: Instant? = null,
    val updatedAt: Instant? = null,
    /**
     * Accumulated execution time, so a cached fallback keeps reporting the runtime's clock
     * instead of reverting to wall clock since [startedAt]. The live anchor is deliberately
     * not cached: a cache fallback is never live, so the value it restores is final.
     */
    val activeDurationMs: Long? = null,
) {
    init {
        require(summary.isNotBlank()) { "summary must not be blank" }
        require(summary.length <= MAX_SUMMARY_CHARS) { "summary exceeds bound" }
    }

    companion object {
        const val MAX_SUMMARY_CHARS: Int = 512
    }
}

fun SkillBillStatusOutcome.toCacheSnapshotOrNull(): LastKnownDisplayCache? =
    when (this) {
        is SkillBillStatusOutcome.Active ->
            LastKnownDisplayCache(
                display = CachedDisplaySnapshot(
                    summary = summary,
                    repositoryIdentity = repositoryIdentity,
                    issueKey = issueKey,
                    currentStepId = currentStepId,
                    currentStepLabel = currentStepLabel,
                    progressCompleted = progressCompleted,
                    progressTotal = progressTotal,
                    startedAt = startedAt,
                    currentSubtaskId = currentSubtaskId,
                    subtaskStartedAt = subtaskStartedAt,
                    updatedAt = updatedAt,
                    activeDurationMs = activeDurationMs,
                ),
                observedAt = observedAt,
            )

        is SkillBillStatusOutcome.Paused ->
            LastKnownDisplayCache(
                display = CachedDisplaySnapshot(
                    summary = summary,
                    repositoryIdentity = repositoryIdentity,
                    issueKey = issueKey,
                    currentStepId = currentStepId,
                    currentStepLabel = currentStepLabel,
                    progressCompleted = progressCompleted,
                    progressTotal = progressTotal,
                    startedAt = startedAt,
                    currentSubtaskId = currentSubtaskId,
                    subtaskStartedAt = subtaskStartedAt,
                    updatedAt = updatedAt,
                    activeDurationMs = activeDurationMs,
                ),
                observedAt = observedAt,
            )

        is SkillBillStatusOutcome.Stale ->
            if (fromCache) {
                null
            } else {
                LastKnownDisplayCache(
                    display = CachedDisplaySnapshot(
                        summary = summary,
                        repositoryIdentity = repositoryIdentity,
                        issueKey = issueKey,
                        currentStepId = currentStepId,
                        currentStepLabel = currentStepLabel,
                        progressCompleted = progressCompleted,
                        progressTotal = progressTotal,
                        startedAt = startedAt,
                        currentSubtaskId = currentSubtaskId,
                        subtaskStartedAt = subtaskStartedAt,
                        updatedAt = updatedAt,
                        activeDurationMs = activeDurationMs,
                    ),
                    observedAt = observedAt,
                )
            }

        is SkillBillStatusOutcome.Blocked ->
            LastKnownDisplayCache(
                display = CachedDisplaySnapshot(
                    summary = summary,
                    repositoryIdentity = repositoryIdentity,
                    issueKey = issueKey,
                    currentStepId = currentStepId,
                    currentStepLabel = currentStepLabel,
                    startedAt = startedAt,
                    currentSubtaskId = currentSubtaskId,
                    subtaskStartedAt = subtaskStartedAt,
                    updatedAt = updatedAt,
                    activeDurationMs = activeDurationMs,
                ),
                observedAt = observedAt,
            )

        is SkillBillStatusOutcome.Failed ->
            LastKnownDisplayCache(
                display = CachedDisplaySnapshot(
                    summary = summary,
                    repositoryIdentity = repositoryIdentity,
                    issueKey = issueKey,
                    currentStepId = currentStepId,
                    currentStepLabel = currentStepLabel,
                    startedAt = startedAt,
                    currentSubtaskId = currentSubtaskId,
                    subtaskStartedAt = subtaskStartedAt,
                    updatedAt = updatedAt,
                    activeDurationMs = activeDurationMs,
                ),
                observedAt = observedAt,
            )

        is SkillBillStatusOutcome.Done ->
            LastKnownDisplayCache(
                display = CachedDisplaySnapshot(
                    summary = summary,
                    repositoryIdentity = repositoryIdentity,
                    issueKey = issueKey,
                    progressCompleted = progressCompleted,
                    progressTotal = progressTotal,
                    startedAt = startedAt,
                    updatedAt = updatedAt,
                    activeDurationMs = activeDurationMs,
                ),
                observedAt = observedAt,
            )

        is SkillBillStatusOutcome.Idle,
        is SkillBillStatusOutcome.Unavailable,
        is SkillBillStatusOutcome.Incompatible,
        -> null
    }

fun LastKnownDisplayCache.toStaleOutcome(): SkillBillStatusOutcome.Stale =
    SkillBillStatusOutcome.Stale(
        observedAt = observedAt,
        summary = display.summary,
        repositoryIdentity = display.repositoryIdentity,
        issueKey = display.issueKey,
        currentStepId = display.currentStepId,
        currentStepLabel = display.currentStepLabel,
        progressCompleted = display.progressCompleted,
        progressTotal = display.progressTotal,
        startedAt = display.startedAt,
        currentSubtaskId = display.currentSubtaskId,
        subtaskStartedAt = display.subtaskStartedAt,
        updatedAt = display.updatedAt,
        activeDurationMs = display.activeDurationMs,
        fromCache = true,
        diagnostic = StatusDiagnostic(reasonCode = "cache_fallback"),
    )

File: vscode-extension/src/domain/LastKnownDisplayCache.ts
import { SkillBillStatusOutcome, StatusDiagnostic } from "./SkillBillStatusOutcome";

export interface CachedDisplaySnapshot {
  summary: string;
  repositoryIdentity?: string;
  issueKey?: string;
  currentStepId?: string;
  currentStepLabel?: string;
  progressCompleted?: number;
  progressTotal?: number;
  startedAt?: Date;
  currentSubtaskId?: string;
  subtaskStartedAt?: Date;
  updatedAt?: Date;
  activeDurationMs?: number;
  subtaskActiveDurationMs?: number;
}

export const MAX_SUMMARY_CHARS = 512;

export interface LastKnownDisplayCache {
  display: CachedDisplaySnapshot;
  observedAt: Date;
}

export function toCacheSnapshotOrNull(outcome: SkillBillStatusOutcome): LastKnownDisplayCache | null {
  switch (outcome.kind) {
    case "active":
      return {
        display: liveSnapshot(outcome),
        observedAt: outcome.observedAt,
      };
    case "paused":
      return {
        display: liveSnapshot(outcome),
        observedAt: outcome.observedAt,
      };
    case "stale":
      if (outcome.fromCache) {
        return null;
      }
      return {
        display: staleSnapshot(outcome),
        observedAt: outcome.observedAt,
      };
    case "blocked":
    case "failed":
      return {
        display: blockedFailedSnapshot(outcome),
        observedAt: outcome.observedAt,
      };
    case "done":
      return {
        display: {
          summary: outcome.summary,
          repositoryIdentity: outcome.repositoryIdentity,
          issueKey: outcome.issueKey,
          progressCompleted: outcome.progressCompleted,
          progressTotal: outcome.progressTotal,
          startedAt: outcome.startedAt,
          updatedAt: outcome.updatedAt,
          activeDurationMs: outcome.activeDurationMs,
        },
        observedAt: outcome.observedAt,
      };
    default:
      return null;
  }
}

function liveSnapshot(
  outcome: Extract<SkillBillStatusOutcome, { kind: "active" | "paused" }>,
): CachedDisplaySnapshot {
  return {
    summary: outcome.summary,
    repositoryIdentity: outcome.repositoryIdentity,
    issueKey: outcome.issueKey,
    currentStepId: outcome.currentStepId,
    currentStepLabel: outcome.currentStepLabel,
    progressCompleted: outcome.progressCompleted,
    progressTotal: outcome.progressTotal,
    startedAt: outcome.startedAt,
    currentSubtaskId: outcome.currentSubtaskId,
    subtaskStartedAt: outcome.subtaskStartedAt,
    updatedAt: outcome.updatedAt,
    activeDurationMs: outcome.activeDurationMs,
    subtaskActiveDurationMs: outcome.subtaskActiveDurationMs,
  };
}

function staleSnapshot(outcome: Extract<SkillBillStatusOutcome, { kind: "stale" }>): CachedDisplaySnapshot {
  return {
    summary: outcome.summary,
    repositoryIdentity: outcome.repositoryIdentity,
    issueKey: outcome.issueKey,
    currentStepId: outcome.currentStepId,
    currentStepLabel: outcome.currentStepLabel,
    progressCompleted: outcome.progressCompleted,
    progressTotal: outcome.progressTotal,
    startedAt: outcome.startedAt,
    currentSubtaskId: outcome.currentSubtaskId,
    subtaskStartedAt: outcome.subtaskStartedAt,
    updatedAt: outcome.updatedAt,
    activeDurationMs: outcome.activeDurationMs,
    subtaskActiveDurationMs: outcome.subtaskActiveDurationMs,
  };
}

function blockedFailedSnapshot(
  outcome: Extract<SkillBillStatusOutcome, { kind: "blocked" | "failed" }>,
): CachedDisplaySnapshot {
  return {
    summary: outcome.summary,
    repositoryIdentity: outcome.repositoryIdentity,
    issueKey: outcome.issueKey,
    currentStepId: outcome.currentStepId,
    currentStepLabel: outcome.currentStepLabel,
    startedAt: outcome.startedAt,
    currentSubtaskId: outcome.currentSubtaskId,
    subtaskStartedAt: outcome.subtaskStartedAt,
    updatedAt: outcome.updatedAt,
    activeDurationMs: outcome.activeDurationMs,
    subtaskActiveDurationMs: outcome.subtaskActiveDurationMs,
  };
}

export function toStaleOutcome(cache: LastKnownDisplayCache): SkillBillStatusOutcome {
  const display = cache.display;
  return {
    kind: "stale",
    observedAt: cache.observedAt,
    summary: display.summary,
    repositoryIdentity: display.repositoryIdentity,
    issueKey: display.issueKey,
    currentStepId: display.currentStepId,
    currentStepLabel: display.currentStepLabel,
    progressCompleted: display.progressCompleted,
    progressTotal: display.progressTotal,
    startedAt: display.startedAt,
    currentSubtaskId: display.currentSubtaskId,
    subtaskStartedAt: display.subtaskStartedAt,
    updatedAt: display.updatedAt,
    activeDurationMs: display.activeDurationMs,
    subtaskActiveDurationMs: display.subtaskActiveDurationMs,
    fromCache: true,
    diagnostic: { reasonCode: "cache_fallback" } satisfies StatusDiagnostic,
  };
}
